pyro/distributions/one_one_matching.py

```diff
@@ -108,20 +108,26 @@ def log_partition_function(self):
         # We find that Sinkhorn iteration is more robust and faster than the
         # optimal belief propagation updates suggested in [1-4].
         finfo = torch.finfo(self.logits.dtype)
-        shift = self.logits.data.max(1, True).values
-        shift.clamp_(min=finfo.min, max=finfo.max)
-        p = (self.logits - shift).exp().clamp(min=finfo.tiny ** 0.5)
-        d = 1 / p.sum(0)
+        # Note gradients are more accurate when shift is not detached.
```

Comment:
whoops i should have pointed this out

Reply:
Oh is this well-known? Mathematically they're identical, and detaching is cheaper. In fact I got a lot of speedup in #2640 by detaching all over the place. I only luckily happened upon not-detaching here and I don't understand why it improves gradient accuracy 🤔

Reply:
oh i guess you're right then i'm confused. should i merge anyway?

Reply:
Yes, ready to merge.

Empirically this has accurate gradients, but I would like to follow up trying to figure out why the gradients are better than the .detach'ed version. Thanks for reviewing!